File: python/leetcode/028.py
class Solution(object):
    def strStr(self, haystack, needle):
        n, m = len(haystack), len(needle)
        for i in range(n - m + 1):
            if haystack[i:i + m] == needle:
                return i
        return -1


class Solution(object):
    def strStr(self, haystack, needle):
        return haystack.find(needle)


class Solution(object):
    def strStr(self, haystack, needle):
        if not needle:
            return 0
        m, next = len(needle), [0] * len(needle)
        for i in range(1, m):
            j = next[i - 1]
            while j > 0 and needle[i] != needle[j]:
                j = next[j - 1]
            if needle[i] == needle[j]:
                next[i] = j + 1
        j = 0
        for i in range(len(haystack)):
            while j > 0 and needle[j] != haystack[i]:
                j = next[j - 1]
            if needle[j] == haystack[i]:
                j += 1
            if j == m:
                return i - m + 1
        return -1

File: cpp/leetcode/028.cpp
class Solution {
public:
    int strStr(string haystack, string needle) {
        int n = haystack.size(), m = needle.size();
        for(int i = 0; i <= n - m; i++){
            if(haystack.substr(i, m) == needle) {
                return i;
            }
        }
        return -1;
    }
};

class Solution {
public:
    int strStr(string haystack, string needle) {
        if (needle.empty()) {
            return 0;
        }
        int m = needle.size();
        vector<int> next(m, 0);
        for (int i = 1; i < m; i++) {
            int j = next[i-1];
            while (j > 0 && needle[i] != needle[j]) {
                j = next[j-1];
            }
            if (needle[i] == needle[j]) {
                next[i] = j + 1;
            }
        }
        for(int i = 0, j = 0; i < haystack.size(); i++){
            while (j > 0 && needle[j] != haystack[i]) {
                j = next[j-1];
            }
            if (needle[j] == haystack[i]) {
                j++;
            }
            if (j == m) {
                return i - m + 1;
            }
        }
        return -1;
    }
};